osquery/events/windows/etw/etw_publisher_processes.cpp

```diff
@@ -0,0 +1,552 @@
+  // 10000000 nanoseconds per second
+  static constexpr LONGLONG expiredTime10secs = 10000000 * 10;
+
+  if (processStartAggregationCache_.empty()) {
+    return;
+  }
+
+  FILETIME currentFiletimeTimestamp{0};
+  GetSystemTimeAsFileTime(&currentFiletimeTimestamp);
+
+  ULARGE_INTEGER currentTimestamp{0};
+  currentTimestamp.HighPart = currentFiletimeTimestamp.dwHighDateTime;
+  currentTimestamp.LowPart = currentFiletimeTimestamp.dwLowDateTime;
+
+  auto it = processStartAggregationCache_.begin();
+  while (it != processStartAggregationCache_.end()) {
+    ULARGE_INTEGER eventTimestamp{0};
+    eventTimestamp.HighPart = it->second->CreateTime.dwHighDateTime;
+    eventTimestamp.LowPart = it->second->CreateTime.dwLowDateTime;
+
+    // check if event is still there after 10 secs
+    if ((eventTimestamp.QuadPart + expiredTime10secs) <
+        currentTimestamp.QuadPart) {
+      // event expire and should be deleted
+      processStartAggregationCache_.erase(it++);
+    } else {
+      ++it;
+    }
+  }
+}
+
+void EtwPublisherProcesses::updateHardVolumeWithLogicalDrive(
+    std::string& path) {
+  // Updating the hardvolume entries with logical volume data
+  for (const auto& [hardVolume, logicalDrive] : hardVolumeDrives_) {
+    size_t pos = 0;
+    if ((pos = path.find(hardVolume, pos)) != std::string::npos) {
+      path.replace(pos, hardVolume.length(), logicalDrive);
+      break;
+    }
+  }
+}
+
+void EtwPublisherProcesses::updateUserInfo(const std::string& userSid,
+                                           std::string& username) {
+  // Updating user information using gathered user SIDs as input
+  auto usernameIt = usernamesBySIDs_.find(userSid);
+  if (usernameIt != usernamesBySIDs_.end()) {
+    auto cachedUsername = usernameIt->second;
+    username.assign(cachedUsername);
+  } else {
+    PSID pSid = nullptr;
+
+    if (!ConvertStringSidToSidA(userSid.c_str(), &pSid) || pSid == nullptr) {
+      // Inserting empty username to avoid the lookup logic to be called again
+      usernamesBySIDs_.insert({userSid, ""});
+      return;
+    }
+
+    std::vector<char> domainNameStr(MAX_PATH - 1, 0x0);
+    std::vector<char> userNameStr(MAX_PATH - 1, 0x0);
+    DWORD domainNameSize = MAX_PATH;
+    DWORD userNameSize = MAX_PATH;
+    SID_NAME_USE sidType = SID_NAME_USE::SidTypeInvalid;
+
+    if (!LookupAccountSidA(NULL,
+                           pSid,
+                           userNameStr.data(),
+                           &userNameSize,
+                           domainNameStr.data(),
+                           &domainNameSize,
+                           &sidType) ||
+        strlen(domainNameStr.data()) == 0 ||
+        strlen(domainNameStr.data()) >= MAX_PATH ||
+        strlen(userNameStr.data()) == 0 ||
+        strlen(userNameStr.data()) >= MAX_PATH ||
+        sidType == SID_NAME_USE::SidTypeInvalid) {
+      // Inserting empty username to avoid the lookup logic to be called again
```

Comment:
I see this comment but I don't actually see the insert. Am I misunderstanding?

Maybe log here as well?

Reply:
Good catch! I've just added the missing insert here. This helper is used to populate an attribute of the resulting event, I'd prefer not to log here